Switch from setting ientry to using nextEvent in a loop

bryngemark opened the pull request:
This fix to `skipToEvent(offset)` switches from setting `ientry_ = offset` to looping using `nextEvent()` `offset` times. Apparently the actual event counter doesn't budge if `ientry_` is just directly updated.

This important bugfix should go in before the real release, preferably along with a fix to randomization of the beam electron position (currently all runs are seeded identically). 

Should close #42

bryngemark committed:
Switch from setting ientry to using nextEvent in a loop

Aparently the actual event counter doesn't budge if ientry is just
directly updated.

bryngemark commented:
Here's a pair of primitive event displays showing before and after, combining the 1e sim events from run 10041 and pileup events from run 10042 in both cases. Pink (filled) is the sim event. 

Red boxes are sim+pileup before the fix (`ientry_` is merely shifted/set), which are on top of pink. 
![overlay_10041-10042_simPink_iEntryShift](https://user-images.githubusercontent.com/42861632/115442806-4ef25400-a1c7-11eb-8e8e-44e9303fdc34.png)
After fix (`ientry_` set using `nextEvent()`): green boxes are sim+pileup, and there are lots of pileup hits outside of the sim event hit locations.
![overlay_10041-10042_simPink_iEntryLoop](https://user-images.githubusercontent.com/42861632/115442810-50238100-a1c7-11eb-8b1c-157832f72977.png)

tomeichlersmith reviewed (approved):
I'm a little shocked that changing `ientry_` didn't work, but the proof is in the pudding :man_shrugging: 

This looks good and since it only happens at the beginning of a run, it's okay if it is a little time-inefficient.

bryngemark commented:
Sooo.... I have figured out that, while setting `ientry_` in the loop helps at first, down the road at the point of event number wrapping, the overlay event counter is no longer getting incremented. I have tried a few things but I'm at a loss here. The figure shows the hit _x_ vs event number, where overlay is black, and sim is pink. You can see that the same single hit from overlay gets repeated starting at event 1022 (which happens to be 10k-8978, which is the random pileup event number offset applied in this case). 

![hitXvsEvNb_TStag_overlayBlack_simPink](https://user-images.githubusercontent.com/42861632/116485554-5dccac80-a840-11eb-8bd2-c03ab0010f35.png)

@tomeichlersmith @omar-moreno  any ideas??

tomeichlersmith commented:
I _feel_ like it's an off-by-one error, so I would try change the -1 to 0.

https://github.com/LDMX-Software/Framework/blob/0338027497d2cd838e7ab91c0d070552363e3846/src/Framework/EventFile.cxx#L203-L209

Besides that... :man_shrugging:

bryngemark commented:
> I _feel_ like it's an off-by-one error, so I would try change the -1 to 0.
> 
> https://github.com/LDMX-Software/Framework/blob/0338027497d2cd838e7ab91c0d070552363e3846/src/Framework/EventFile.cxx#L203-L209
> 
> Besides that... 🤷‍♂️

Yeah I tried that. I also checked that the event moves through `nextEvent()` as expected after resetting `ientry_`. I tried changing the event number in the event header to start from 0 after running `skipToEvent` but that doesn't matter.

I think there's something profound about how to set `ientry_` that I'm missing. That it didn't work to just set, but we had to loop using `nextEvent()` makes me suspicious.

bryngemark commented:
If we really don't think we can figure this out, I can work around this for now (outside of `EventFile.cxx`) by entirely avoiding the event number wrapping as follows:
1. draw a pileup start event number from `(2,X)` where `X` is rather small
2. set `p.maxEvents = N_pileup_events - X` events 

I've done this successfully for a few files to get TS validation unstuck. At least my checks indicate that this results in the desired distributions. It's patchy and if we go that route, I'd like to keep track of this problem to solve it for real later.

For large-scale production, we anticipate reusing every pileup file ~10 times.`X=100` should probably be enough to avoid more than occasionally reusing the exact same sequence of pileup events on the (anyway random) sim events. (With proper beam electron position randomization, that is.)

An alternative to step 2 would be to generate the pileup files with an extra `X` events in them and use this as a buffer for the start event offset. If we can't use event wrapping, I would anyway have to generate enough pileup events per file to cover the fluctuations in surviving PN events per 1M file.

tomeichlersmith commented:
Another option (lame and crude but would probably work) is to close and re-open the pileup file in the OverlayProducer. Something like

```cpp
// in produce
if (!overlayFile_->nextEvent()) {
  // we reached the end of the overlay file
  //  restart by closing the file and re-opening it at the beginning.
  overlayFile_->close(); // closes TFile
  overlayEvent_->onEndOfFile(); // detach TTree addresses
  overlayFile_ =
      std::make_unique<framework::EventFile>(params_, overlayFileName_, true);
  overlayFile_->setupEvent(&overlayEvent_);
}
```

Actually, _this_ could be why the looping isn't working... Maybe adding a reset to the event bus in the loop-able part of the tree would do it.
```cpp
if (event_) {
  event_->onEndOfFile();
  this->setupEvent(event_);
}
```

bryngemark commented:
> Another option (lame and crude but would probably work) is to close and re-open the pileup file in the OverlayProducer. Something like
> 
> ```c++
> // in produce
> if (!overlayFile_->nextEvent()) {
>   // we reached the end of the overlay file
>   //  restart by closing the file and re-opening it at the beginning.
>   overlayFile_->close(); // closes TFile
>   overlayEvent_->onEndOfFile(); // detach TTree addresses
>   overlayFile_ =
>       std::make_unique<framework::EventFile>(params_, overlayFileName_, true);
>   overlayFile_->setupEvent(&overlayEvent_);
> }
> ```
>

mh indeed it _is_ lame... but probably a working last resort. let's keep it in our back pockets.
 
> Actually, _this_ could be why the looping isn't working... Maybe adding a reset to the event bus in the loop-able part of the tree would do it.
> 
> ```c++
> if (event_) {
>   event_->onEndOfFile();
>   this->setupEvent(event_);
> }
> ```

ah, right, yes i had wanted to try `setupEvent()` but i forgot. i'll play with this.

bryngemark committed:
Fix event wrapping

Make sure event is really reset, instead of stuck on last one before
wrapping at the end of the file

bryngemark commented:
> > Actually, _this_ could be why the looping isn't working... Maybe adding a reset to the event bus in the loop-able part of the tree would do it.
> > ```c++
> > if (event_) {
> >   event_->onEndOfFile();
> >   this->setupEvent(event_);
> > }
> > ```
> 
> ah, right, yes i had wanted to try `setupEvent()` but i forgot. i'll play with this.

YES this fixed it. 

![overlay_10041-10042_simPink_iEntryReset](https://user-images.githubusercontent.com/42861632/116629212-94202f80-a905-11eb-8040-e714f49339f1.png)

tomeichlersmith reviewed (approved):
Glad to hear it is working :smile:

tomeichlersmith commented:
Using my awesome administrator powers :zap: :zap: I will merge this PR so that I can move the GitHub actions PR forward.
